### PR TITLE
evidence: improve amnesia evidence handling

### PR DESCRIPTION
**Pull request by cmwaters**
## Description

Added a bit more rigor to the testing and whilst doing that uncovered a few overlooked bugs. 

- PotentialAmnesiaEvidence because it gets stored in a different bucket wasn't being added to the evidenceList and hence wasn't being gossiped

- Inbound amnesia evidence wasn't being handled correctly i.e inbound amnesia evidence with an empty polc shouldn't be added but the potential amnesia evidence part should be extracted and the node begin it's own trial period

- There wasn't a method to check if an inbound potential evidence is already on trial but now there is

- The height with which once reached we upgrade potential evidence to amnesia evidence wasn't being properly updated. 

- Evidence was inconsistently switching between concrete types and pointers. Now it's been changed to using just pointers. 

I also added a lot more information to the logger so it is easier to understand what the evidence module is doing and I added a bit more documentation

### CODE REVIEW

File: evidence/doc.go [hunk 1]
```diff
@@ -24,15 +24,39 @@ uncommitted evidence at intervals of 60 seconds (set by the by broadcastEvidence
 It uses a concurrent list to store the evidence and before sending verifies that each evidence is still valid in the
 sense that it has not exceeded the max evidence age and height (see types/params.go#EvidenceParams).
 
+Three are four buckets that evidence can be stored in: Pending, Committed, Awaiting and POLC's.
+
+1. Pending is awaiting to be committed (evidence is usually broadcasted then)
+
+2. Committed is for those already on the block and is to ensure that evidence isn't submitted twice
+
+3. Awaiting primarily refers to PotentialAmnesiaEvidence which must wait for a trial period before
```
Comment (melekes) [inline evidence/doc.go:33]: maybe rename to awaiting_trial

File: evidence/pool.go [hunk 2]
```diff
@@ -206,68 +210,36 @@ func (evpool *Pool) AddEvidence(evidence types.Evidence) error {
...
+					PotentialAmnesiaEvidence: ae.PotentialAmnesiaEvidence,
+					Polc:                     types.EmptyPOLC(),
 				}
-				key := keyAwaiting(pe)
-				err = evpool.evidenceStore.Set(key, evBytes)
-				if err != nil {
-					return err
+				if evpool.IsPending(aeWithoutPolc) {
+					evpool.removePendingEvidence(aeWithoutPolc)
 				}
```
Comment (melekes) [inline evidence/pool.go:235]: else if ?

Comment (cmwaters) [reply evidence/pool.go:235]: yeah technically they shouldn't both happen

**Comment by cmwaters** [comment]
Have cleaned up all evidence so that the different types all just use pointers

File: evidence/pool.go [hunk 3]
```diff
@@ -157,6 +161,8 @@ func (evpool *Pool) AddEvidence(evidence types.Evidence) error {
 		evList = []types.Evidence{evidence}
 	)
 
+	evpool.logger.Debug("Attempting to add evidence", "evidence", evidence)
```
Comment (melekes) [inline evidence/pool.go:164]:
```suggestion
	evpool.logger.Debug("Attempting to add evidence", "ev", evidence)
```

File: evidence/pool.go [hunk 4]
```diff
@@ -187,8 +193,12 @@ func (evpool *Pool) AddEvidence(evidence types.Evidence) error {
 	}
 
 	for _, ev := range evList {
+
 		if evpool.Has(ev) {
-			continue
+			// if it is an amnesia evidence we have but
```
Comment (melekes) [inline evidence/pool.go:198]:
```suggestion
			// if it is an amnesia evidence we have, but POLC is absent
```

File: evidence/pool.go [hunk 5]
```diff
@@ -205,69 +215,39 @@ func (evpool *Pool) AddEvidence(evidence types.Evidence) error {
 			return fmt.Errorf("failed to verify %v: %w", ev, err)
 		}
 
+		evpool.logger.Info("Hi")
```
Comment (melekes) [inline evidence/pool.go:218]: ???

File: evidence/pool.go [hunk 6]
```diff
@@ -205,69 +215,39 @@ func (evpool *Pool) AddEvidence(evidence types.Evidence) error {
...
-				pbe, err := types.EvidenceToProto(pe)
-				if err != nil {
-					return err
-				}
-				evBytes, err := pbe.Marshal()
-				if err != nil {
-					return err
+			continue
+		} else if ae, ok := ev.(*types.AmnesiaEvidence); ok {
+			evpool.logger.Info("Hi")
```
Comment (melekes) [inline evidence/pool.go:227]: ???

File: evidence/pool.go [hunk 7]
```diff
@@ -205,69 +215,39 @@ func (evpool *Pool) AddEvidence(evidence types.Evidence) error {
...
-						evpool.logger.Error("Failed to create amnesia evidence from potential amnesia evidence", "err", err)
-						// revert back to processing potential amnesia evidence
-						exists = false
-					} else {
-						evpool.logger.Info("Formed amnesia evidence from own polc", "amnesiaEvidence", ae)
-					}
-					break
-				}
+		// to form AmensiaEvidence else start the trial period for the piece of evidence
+		if pe, ok := ev.(*types.PotentialAmnesiaEvidence); ok {
```
Comment (melekes) [inline evidence/pool.go:221]: `switch ev.(type)` might be a good fit here

File: evidence/pool.go [hunk 8]
```diff
@@ -205,69 +215,39 @@ func (evpool *Pool) AddEvidence(evidence types.Evidence) error {
...
 				}
-				key := keyAwaiting(pe)
-				err = evpool.evidenceStore.Set(key, evBytes)
-				if err != nil {
-					return err
+				continue
+			} else {
+				// we are going to add this amnesia evidence and check if we already have an amnesia evidence or potential
+				// amnesia evidence that addesses the same case
+				evpool.logger.Info("Hi")
```
Comment (melekes) [inline evidence/pool.go:237]: ???

File: evidence/pool.go [hunk 9]
```diff
@@ -359,31 +339,46 @@ func (evpool *Pool) IsPending(evidence types.Evidence) bool {
 	return ok
 }
 
+// IsOnTrial checks whether a piece of evidence is in the awaiting bucket.
+// Only Potential Amnesia Evidence is stored here.
+func (evpool *Pool) IsOnTrial(evidence types.Evidence) bool {
+	if pe, ok := evidence.(*types.PotentialAmnesiaEvidence); ok {
```
Comment (melekes) [inline evidence/pool.go:345]: early return ...

```suggestion
	pe, ok := evidence.(*types.PotentialAmnesiaEvidence)
	if !ok {
	  return false
	}
```

File: evidence/pool.go [hunk 10]
```diff
@@ -631,6 +634,81 @@ func (evpool *Pool) upgradePotentialAmnesiaEvidence() int64 {
...
+		evBytes, err := pbe.Marshal()
+		if err != nil {
+			return err
+		}
+		key := keyAwaitingTrial(pe)
+		err = evpool.evidenceStore.Set(key, evBytes)
+		if err != nil {
+			return err
+		}
+		evpool.logger.Debug("Valid potential amnesia evidence has been added. Starting trial period.",
```
Comment (melekes) [inline evidence/pool.go:698]:
```suggestion
		evpool.logger.Debug("Valid potential amnesia evidence has been added. Starting trial period",
```

File: evidence/pool.go [hunk 11]
```diff
@@ -631,6 +634,81 @@ func (evpool *Pool) upgradePotentialAmnesiaEvidence() int64 {
...
+			"PotentialAmnesiaEvidence", pe)
```
Comment (melekes) [inline evidence/pool.go:699]:
```suggestion
			"ev", pe)
```

File: types/evidence.go [hunk 12]
```diff
@@ -872,27 +809,26 @@ func (e PhantomValidatorEvidence) Verify(chainID string, pubKey crypto.PubKey) e
...
-		return false
 	}
+
+	return false
 }
 
-func (e PhantomValidatorEvidence) ValidateBasic() error {
+func (e *PhantomValidatorEvidence) ValidateBasic() error {
+	if e == nil {
+		return errors.New("emoty phantom validator evidence")
```
Comment (melekes) [inline types/evidence.go:823]:
```suggestion
		return errors.New("empty phantom validator evidence")
```

File: evidence/pool.go (hunk 6 above)
Comment (cmwaters) [reply evidence/pool.go:227]: Okay lol this was all part of my debugging which I thought I had removed afterwards but turns out I was pretty slack

File: types/evidence.go [hunk 13]
```diff
@@ -768,6 +764,15 @@ type PhantomValidatorEvidence struct {
 
 var _ Evidence = &PhantomValidatorEvidence{}
 
+// NewPhantomValidatorEvidence creates a new instance of the respective evidence 
+func NewPhantomValidatorEvidence(vote *Vote, lastHeightValidatorWasInSet int64) *PhantomValidatorEvidence {
+	if vote == nil || lastHeightValidatorWasInSet <= 0 { return nil }
```
Comment (melekes) [inline types/evidence.go]: why are u duplicating ValidateBasic logic here?

Comment (cmwaters) [reply types/evidence.go]: I was just replicating what DuplicateVoteEvidence does - which is add null checks

Comment (melekes) [reply types/evidence.go]: it hides errors. that's not a good style. if anything, the implementation should panic or return an error if someone calls a constructor w/ nil vote

Comment (melekes) [reply types/evidence.go]: I'd remove the check

Comment (cmwaters) [reply types/evidence.go]: okay. I also thought about having the option to return an error or even to validate basic when we create the evidence

Comment (cmwaters) [reply types/evidence.go]: but I will remove the checks on all of them except the constructors for DuplicateVoteEvidence and PotentialAmnesiaEvidence which need to order the votes correctly

File: types/evidence.go [hunk 14]
```diff
@@ -875,27 +820,26 @@ func (e PhantomValidatorEvidence) Verify(chainID string, pubKey crypto.PubKey) e
...
+		return errors.New("empty phantom validator evidence")
+	}
 
 	if e.Vote == nil {
 		return errors.New("empty vote")
 	}
 
 	if err := e.Vote.ValidateBasic(); err != nil {
-		return fmt.Errorf("invalid signature: %v", err)
+		return fmt.Errorf("invalid block: %v", err)
```
Comment (marbar3778) [inline types/evidence.go:842]:
```suggestion
		return fmt.Errorf("invalid vote: %W", err)
```
?

File: types/evidence_test.go [hunk 15]
```diff
@@ -638,58 +660,55 @@ func TestEvidenceProto(t *testing.T) {
...
-		{"ConflictingHeadersEvidence empty fail", ConflictingHeadersEvidence{}, false, true},
-		{"ConflictingHeadersEvidence nil H2", ConflictingHeadersEvidence{H1: h1, H2: nil}, false, true},
-		{"ConflictingHeadersEvidence nil H1", ConflictingHeadersEvidence{H1: nil, H2: h2}, false, true},
-		{"ConflictingHeadersEvidence success", ConflictingHeadersEvidence{H1: h1, H2: h2}, false, false},
-		{"LunaticValidatorEvidence empty fail", LunaticValidatorEvidence{}, false, true},
-		{"LunaticValidatorEvidence only header fail", LunaticValidatorEvidence{Header: header1}, false, true},
-		{"LunaticValidatorEvidence only vote fail", LunaticValidatorEvidence{Vote: v}, false, true},
-		{"LunaticValidatorEvidence header & vote fail", LunaticValidatorEvidence{Header: header1, Vote: v}, false, true},
-		{"LunaticValidatorEvidence success", LunaticValidatorEvidence{Header: header1,
+		{"DuplicateVoteEvidence empty fail", &DuplicateVoteEvidence{}, false, true},
```
Comment (marbar3778) [inline types/evidence_test.go:668]: could you add a nil case 
```suggestion
        {"nil fail", nil, true, true},
		{"DuplicateVoteEvidence empty fail", &DuplicateVoteEvidence{}, false, true},
```

File: types/evidence.go (hunk 14 above)
Comment (melekes) [reply types/evidence.go]: `%w`

File: types/block.go [hunk 16]
```diff
@@ -90,11 +90,11 @@ func (b *Block) ValidateBasic() error {
 	// NOTE: b.Evidence.Evidence may be nil, but we're just looping.
 	for i, ev := range b.Evidence.Evidence {
 		switch ev.(type) {
-		case *ConflictingHeadersEvidence, ConflictingHeadersEvidence:
+		case *ConflictingHeadersEvidence:
 			// ConflictingHeadersEvidence must be broken up in pieces and never
 			// committed as a single piece.
 			return fmt.Errorf("found ConflictingHeadersEvidence (#%d)", i)
-		case *PotentialAmnesiaEvidence, PotentialAmnesiaEvidence:
```
Comment (melekes) [inline types/block.go:97]: I'd prefer for these to stay just in case

Comment (cmwaters) [reply types/block.go:97]: I was getting these errors: 
```
types/block.go:93:3: impossible type switch case: ev (type Evidence) cannot have dynamic type ConflictingHeadersEvidence (Address method has pointer receiver)
types/block.go:97:3: impossible type switch case: ev (type Evidence) cannot have dynamic type PotentialAmnesiaEvidence (Address method has pointer receiver)
```

**Pull request merged**